### PR TITLE
Extract common PlatformView functionality: Gesture and PointerEvent

### PR DESCRIPTION
**Pull request by cyanglaz**
## Description

PlatformViewSurface handling gesture and pointer events

## Related Issues

#36779

## Tests

I added the following tests:

* PlatformViewSurface view can lose gesture arenas
* PlatformViewSurface gesture recognizers dispatch events
* PlatformViewSurface can claim gesture after all pointers are up
* PlatformViewSurface rebuilt during gesture
* PlatformViewSurface with eager gesture recognizer
* PlatformViewRenderBox reconstructed with same gestureRecognizers
* PlatformViewSurface rebuilt with same gestureRecognizers

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/rendering/platform_view.dart [hunk 1]
```diff
@@ -759,8 +764,54 @@ class PlatformViewRenderBox extends RenderBox {
     }
   }
 
+  /// How to behave during hit testing.
+  // The implicit setter is enough here as changing this value will just affect
+  // any newly arriving events there's nothing we need to invalidate.
+  PlatformViewHitTestBehavior hitTestBehavior;
+
+  /// {@template flutter.rendering.platformView.updateGestureRecognizers}
```
Comment (amirh) [inline packages/flutter/lib/src/rendering/platform_view.dart]: /// {@macro flutter.rendering.platformView.updateGestureRecognizers}

File: packages/flutter/lib/src/rendering/platform_view.dart [hunk 2]
```diff
@@ -790,4 +849,113 @@ class PlatformViewRenderBox extends RenderBox {
...
+  }
+}
+
+// This recognizer constructs gesture recognizers from a set of gesture recognizer factories
+// it was given, adds all of them to a gesture arena team with the _PlatformViewGestureRecognizerCaptain
+// as the team captain.
+// As long as the gesture arena is unresolved the recognizer caches all pointer events.
+// When the team wins the recognizer sends all the cached point events to the [PlatformViewController], and
+// sets itself to a "forwarding mode" where it will forward any new pointer event to the [PlatformViewController].
+class _PlatformViewGestureRecognizerCaptain extends OneSequenceGestureRecognizer {
```
Comment (amirh) [inline packages/flutter/lib/src/rendering/platform_view.dart]: This is almost identical to the existing _AndroidViewGestureRecognizer, the only difference is `dispatcher.handlePointerEvent` vs. `controller.dispatchPointerEvent`.

How about we make the original implementation take a `_PointerEventDispatcher` and use it for both RenderAndroidView and RenderPlatformView ?

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 3]
```diff
@@ -603,22 +603,72 @@ class PlatformViewSurface extends LeafRenderObjectWidget {
...
   ///
   /// [PlatformViewController] is used for dispatching touch events to the platform view.
   /// [PlatformViewController.viewId] identifies the platform view whose contents are painted by this widget.
   final PlatformViewController controller;
 
+  /// Which gestures should be forwarded to the PlatformView.
+  ///
+  /// {@macro flutter.widgets.platformViews.gestureRecognizersDescHead}
+  ///
+  /// For example, with the following setup vertical drags will not be dispatched to the UIKit
```
Comment (amirh) [inline packages/flutter/lib/src/widgets/platform_view.dart]: platform view

**Comment by cyanglaz** [comment]
Updated with comments. Ready for another review.

File: packages/flutter/lib/src/rendering/platform_view.dart [hunk 4]
```diff
@@ -486,15 +486,17 @@ class _UiKitViewGestureRecognizer extends OneSequenceGestureRecognizer {
...
-// As long as ta gesture arena is unresolved the recognizer caches all pointer events.
-// When the team wins the recognizer sends all the cached point events to the embedded Android view, and
-// sets itself to a "forwarding mode" where it will forward any new pointer event to the Android view.
-class _AndroidViewGestureRecognizer extends OneSequenceGestureRecognizer {
-  _AndroidViewGestureRecognizer(
-    this.dispatcher,
+// As long as the gesture arena is unresolved, the recognizer caches all pointer events.
+// When the team wins, the recognizer sends all the cached point events to `_handlePointerEvent`, and
+// sets itself to a "forwarding mode" where it will forward any new pointer event to `_handlePointerEvent`.
+class _PlatformViewGestureRecognizerCaptain extends OneSequenceGestureRecognizer {
```
Comment (amirh) [inline packages/flutter/lib/src/rendering/platform_view.dart]: _PlatformViewGestureRecognizer

File: packages/flutter/lib/src/rendering/platform_view.dart [hunk 5]
```diff
@@ -505,18 +507,19 @@ class _AndroidViewGestureRecognizer extends OneSequenceGestureRecognizer {
...
   // Maps a pointer to a list of its cached pointer events.
   // Before the arena for a pointer is resolved all events are cached here, if we win the arena
-  // the cached events are dispatched to the view, if we lose the arena we clear the cache for
+  // the cached events are dispatched to `_handlePointerEvent`, if we lose the arena we clear the cache for
   // the pointer.
   final Map<int, List<PointerEvent>> cachedEvents = <int, List<PointerEvent>>{};
 
   // Pointer for which we have already won the arena, events for pointers in this set are
-  // immediately dispatched to the Android view.
+  // immediately dispatched to the `_handlePointerEvent`.
```
Comment (amirh) [inline packages/flutter/lib/src/rendering/platform_view.dart]: nit: s/the//

File: packages/flutter/lib/src/rendering/platform_view.dart [hunk 6]
```diff
@@ -486,15 +486,17 @@ class _UiKitViewGestureRecognizer extends OneSequenceGestureRecognizer {
...
+// it was give, adds all of them to a gesture arena team with the _PlatformViewGestureRecognizerCaptain
 // as the team captain.
-// As long as ta gesture arena is unresolved the recognizer caches all pointer events.
-// When the team wins the recognizer sends all the cached point events to the embedded Android view, and
-// sets itself to a "forwarding mode" where it will forward any new pointer event to the Android view.
-class _AndroidViewGestureRecognizer extends OneSequenceGestureRecognizer {
-  _AndroidViewGestureRecognizer(
-    this.dispatcher,
+// As long as the gesture arena is unresolved, the recognizer caches all pointer events.
+// When the team wins, the recognizer sends all the cached point events to `_handlePointerEvent`, and
```
Comment (amirh) [inline packages/flutter/lib/src/rendering/platform_view.dart]: nit: s/point/pointer/

File: packages/flutter/lib/src/rendering/platform_view.dart [hunk 7]
```diff
@@ -759,8 +767,43 @@ class PlatformViewRenderBox extends RenderBox {
...
+  /// How to behave during hit testing.
+  // The implicit setter is enough here as changing this value will just affect
+  // any newly arriving events there's nothing we need to invalidate.
+  PlatformViewHitTestBehavior hitTestBehavior;
+
+  /// {@macro  flutter.rendering.platformView.updateGestureRecognizers}
+  ///
+  /// Any active gesture arena the `PlatformView` participates in is rejected when the
+  /// set of gesture recognizers is changed.
+  void updateGestureRecognizers(Set<Factory<OneSequenceGestureRecognizer>> gestureRecognizers) {
```
Comment (amirh) [inline packages/flutter/lib/src/rendering/platform_view.dart]: What do you think about extracting all the common touch logic for PlatformViewRenderBox and AndroidViewRenderBox to a mixin?

**Comment by cyanglaz** [comment]
Updated.
Note: With the mixin, we are going to move the hitTestBehavior to the mixin out from the RenderAndroidView.
Do we consider this kind of change a breaking change?

File: packages/flutter/lib/src/rendering/platform_view.dart [hunk 8]
```diff
@@ -791,3 +756,65 @@ class PlatformViewRenderBox extends RenderBox {
     config.platformViewId = _controller.viewId;
   }
 }
+
+/// The Mixin handling the pointer events and gestures of a platform view render box.
+mixin PlatformViewGestureMixin on RenderBox {
```
Comment (amirh) [inline packages/flutter/lib/src/rendering/platform_view.dart]: make this private

File: packages/flutter/lib/src/rendering/platform_view.dart [hunk 9]
```diff
@@ -791,3 +756,65 @@ class PlatformViewRenderBox extends RenderBox {
...
+    }
+    result.add(BoxHitTestEntry(this, position));
+    return hitTestBehavior == PlatformViewHitTestBehavior.opaque;
+  }
+
+  @override
+  bool hitTestSelf(Offset position) => hitTestBehavior != PlatformViewHitTestBehavior.transparent;
+
+  @override
+  bool get sizedByParent => true;
```
Comment (amirh) [inline packages/flutter/lib/src/rendering/platform_view.dart]: These aren't related to gesture integration, probably shouldn't be in this mixin.

File: packages/flutter/lib/src/rendering/platform_view.dart [hunk 10]
```diff
@@ -791,3 +756,65 @@ class PlatformViewRenderBox extends RenderBox {
...
+  /// How to behave during hit testing.
+  // The implicit setter is enough here as changing this value will just affect
+  // any newly arriving events there's nothing we need to invalidate.
+  PlatformViewHitTestBehavior hitTestBehavior;
+
+  /// {@macro  flutter.rendering.platformView.updateGestureRecognizers}
+  ///
+  /// Any active gesture arena the `PlatformView` participates in is rejected when the
+  /// set of gesture recognizers is changed.
+  void updateGestureRecognizersWithCallBack(Set<Factory<OneSequenceGestureRecognizer>> gestureRecognizers, _HandlePointerEvent _handlePointerEvent) {
```
Comment (amirh) [inline packages/flutter/lib/src/rendering/platform_view.dart]: Make this private

**Review by amirh — APPROVED** [review]
LGTM

**Pull request merged**